Bluetooth: Mesh: Prevent CDB reusing address of deleted nodes

PavelVPV opened the pull request:
According to MshPRFv1.0.1, section 3.10.7, the Provisioner shall only
use reuse addresses of the deleted nodes after IV Index is updated.

This change prevents CDB reusing addresses of the deleted nodes until IV
Index is updated. Due to a high range of the unicast addresses, CDB only
stores the highest deleted address once a node is deleted. This creates
a limitation where allocating a node with the high primary unicast
address and then deleting it will make CDB skip big range of address
until next IV Index update.

PavelVPV committed:
Bluetooth: Mesh: Prevent CDB reusing address of deleted nodes

According to MshPRFv1.0.1, section 3.10.7, the Provisioner shall only
use reuse addresses of the deleted nodes after IV Index is updated.

This change prevents CDB reusing addresses of the deleted nodes until IV
Index is updated. Due to a high range of the unicast addresses, CDB only
stores the highest deleted address once a node is deleted. This creates
a limitation where allocating a node with the high primary unicast
address and then deleting it will make CDB skip big range of address
until next IV Index update.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

ludvigsj commented on subsys/bluetooth/mesh/cdb.c:
```diff
@@ -855,6 +867,11 @@ void bt_mesh_cdb_node_del(struct bt_mesh_cdb_node *node, bool store)
 	if (IS_ENABLED(CONFIG_BT_SETTINGS) && store) {
 		update_cdb_node_settings(node, false);
+
+		if (node->addr + node->num_elem > bt_mesh_cdb.lowest_avail_addr) {
```
The in-memory increment should not be in the `if (IS_ENABLED(CONFIG_BT_SETTINGS) && store)` right?

PavelVPV replied:
I think if the node is not removed from the persistent storage, then it will appear again after next reboot. If I'm not mistaken, the use case for `store` option is when a node was created but not stored in the persistent storage via `bt_mesh_cdb_node_store`.

ludvigsj replied:
It will reappear on the next reboot, but in the meantime the lowest_avail_addr will cause the address to be available for reuse, no? Meaning
- Create node 0x200 without storing
- Node 0x200 does messaging and starts appearing in RPLs
- `bt_mesh_cdb_node_del(&node_200, false)` (or with persistence off)
- [Reassignment possible here]
- Reboot

PavelVPV replied:
After the first step the node must be stored, otherwise it is an incorrect use of the API. Whenever you allocate a node that gets provisioned to the network -> store it. `bt_mesh_cdb_node_del` with `false` argument is called in 2 cases from what I see:
1. When was not restored from settings;
2. When provisioner didn't finish a node provisioning and was reset;

ludvigsj replied:
Is it not possible to use the provisioner/CDB with `CONFIG_BT_SETTINGS=n`? Why all the conditionals on that throughout then?

PavelVPV replied:
It is possible. Why to set `CONFIG_BT_SETTINGS=n` is a different question I think, but users can do this.

ludvigsj replied:
And then they would be able to do this colission case I described above, right?

PavelVPV replied:
Alright, I see your point.

ludvigsj commented on include/zephyr/bluetooth/mesh/cdb.h:
```diff
@@ -123,14 +124,19 @@ void bt_mesh_cdb_iv_update(uint32_t iv_index, bool iv_update);
- *  @return The new node or NULL if it cannot be allocated.
+ *  @return The new node or NULL if CDB has already allocated @ref CONFIG_BT_MESH_CDB_NODE_COUNT
+ *  nodes, or reached the end of the unicast address range, or if @c addr is non-zero and less than
+ *  the lowest available address.
```
Since you expanded on this: also returns `NULL` for a colliding address even if that address is > lowest available address.

jhedberg commented on include/zephyr/bluetooth/mesh/cdb.h:
```diff
@@ -123,14 +124,20 @@ void bt_mesh_cdb_iv_update(uint32_t iv_index, bool iv_update);
- *  @return The new node or NULL if it cannot be allocated.
+ *  @return The new node or NULL if CDB has already allocated
+ *  :kconfig:option:`CONFIG_BT_MESH_CDB_NODE_COUNT` nodes, or reached the end of the unicast
+ *  address range, or if @c addr is non-zero and less than the lowest available address or collide
```
Could you use the same line length (presumably 80) as the existing documentation in this file? It looks a bit odd that you’re suddenly splitting lines quite differently than what’s already here. Generally, even though the Zephyr requirement has changed for 80 to 100, longer lines should be reserved to code only when it makes it more readable compared to splitting before 80.

jhedberg commented on subsys/bluetooth/mesh/cdb.c:
```diff
@@ -64,10 +64,11 @@ struct app_key_val {
 struct net_val {
 	uint32_t iv_index;
 	bool  iv_update;
+	uint16_t lowest_avail_addr;
```
How are you dealing with stored values written to flash using prior versions with the smaller struct size? It looks like the cdb_net_set() function would simply fail to restore anything in that case. Is that what we want and the best we can do?

PavelVPV replied:
They will be dropped. CDB is still marked as experimental, that is why I though changing the struct is OK.

PavelVPV replied:
Updated to try recovering previous struct. Please have a look.

carlescufi commented:
@PavelVPV please rebase

PavelVPV committed:
Bluetooth: Mesh: Fix print of newly allocated node in CDB

Using the passed address is incorrect since the passed address can be
zero to find the lowest available address.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
@jhedberg @ludvigsj please re-approve after rebase.

PavelVPV commented:
@jhedberg could you please re-approve?
